File: 351002/Lashchenka/src/main/java/by/bsuir/distcomp/service/TweetService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.TweetRequestTo;
import by.bsuir.distcomp.dto.response.TweetResponseTo;
import by.bsuir.distcomp.entity.Tweet;
import by.bsuir.distcomp.exception.DuplicateException;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.TweetMapper;
import by.bsuir.distcomp.repository.impl.InMemoryEditorRepository;
import by.bsuir.distcomp.repository.impl.InMemoryTweetRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TweetService {

    private final InMemoryTweetRepository tweetRepository;
    private final InMemoryEditorRepository editorRepository;
    private final TweetMapper tweetMapper;

    public TweetService(InMemoryTweetRepository tweetRepository, 
                        InMemoryEditorRepository editorRepository,
                        TweetMapper tweetMapper) {
        this.tweetRepository = tweetRepository;
        this.editorRepository = editorRepository;
        this.tweetMapper = tweetMapper;
    }

    public TweetResponseTo create(TweetRequestTo dto) {
        if (!editorRepository.existsById(dto.getEditorId())) {
            throw new ResourceNotFoundException("Editor with id " + dto.getEditorId() + " not found", 40404);
        }
        if (tweetRepository.existsByTitle(dto.getTitle())) {
            throw new DuplicateException("Tweet with title '" + dto.getTitle() + "' already exists", 40303);
        }
        Tweet entity = tweetMapper.toEntity(dto);
        entity.setCreated(LocalDateTime.now());
        entity.setModified(LocalDateTime.now());
        Tweet saved = tweetRepository.save(entity);
        return tweetMapper.toResponseDto(saved);
    }

    public TweetResponseTo getById(Long id) {
        Tweet entity = tweetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + id + " not found", 40405));
        return tweetMapper.toResponseDto(entity);
    }

    public List<TweetResponseTo> getAll() {
        return tweetRepository.findAll().stream()
                .map(tweetMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public TweetResponseTo update(TweetRequestTo dto) {
        Tweet existingTweet = tweetRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + dto.getId() + " not found", 40406));
        if (!editorRepository.existsById(dto.getEditorId())) {
            throw new ResourceNotFoundException("Editor with id " + dto.getEditorId() + " not found", 40407);
        }
        if (tweetRepository.existsByTitleAndIdNot(dto.getTitle(), dto.getId())) {
            throw new DuplicateException("Tweet with title '" + dto.getTitle() + "' already exists", 40304);
        }
        Tweet entity = tweetMapper.toEntity(dto);
        entity.setCreated(existingTweet.getCreated());
        entity.setModified(LocalDateTime.now());
        Tweet updated = tweetRepository.update(entity);
        return tweetMapper.toResponseDto(updated);
    }

    public void deleteById(Long id) {
        if (!tweetRepository.existsById(id)) {
            throw new ResourceNotFoundException("Tweet with id " + id + " not found", 40408);
        }
        tweetRepository.deleteById(id);
    }
}

File: 351003/Kolbeko/app/services/tweet_service.py
from datetime import datetime
from app.repository.memory import tweet_repo, author_repo
from app.schemas.tweet import TweetRequestTo, TweetResponseTo
from app.core.exceptions import AppException

class TweetService:
    def create(self, dto: TweetRequestTo) -> TweetResponseTo:
        if not author_repo.find_by_id(dto.authorId):
            raise AppException(400, "Author not found", 4)
        now = datetime.now().isoformat()
        data = dto.model_dump()
        data.update({"created": now, "modified": now})
        return TweetResponseTo(**tweet_repo.save(data))

    def get_all(self):
        return [TweetResponseTo(**t) for t in tweet_repo.find_all()]

    def get_by_id(self, id: int):
        res = tweet_repo.find_by_id(id)
        if not res: raise AppException(404, "Tweet not found", 5)
        return TweetResponseTo(**res)

    def update(self, id: int, dto: TweetRequestTo) -> TweetResponseTo:
        if not author_repo.find_by_id(dto.authorId):
            raise AppException(400, "Author not found", 7)
        
        existing = tweet_repo.find_by_id(id)
        now = datetime.now().isoformat()
        data = dto.model_dump()
        data["id"] = id
        
        if existing:
            data["created"] = existing.get("created", now)
            data["modified"] = now
        else:
            data["created"] = now
            data["modified"] = now
            
        return TweetResponseTo(**tweet_repo.save(data))

    def delete(self, id: int):
        if not tweet_repo.delete(id):
            raise AppException(404, "Tweet not found", 8)